### PR TITLE
BUG/Perf: Support ExtensionArrays in where

### LINKED ISSUE #16983: BUG: DataFrame.astype with category dtype
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
df = pd.DataFrame({'a': list('aab'), 'b': list('bba')})
df.a = df.a.astype('category')
df.b = df.b.astype('category')
df.astype('object')
# ValueError: Wrong number of dimensions
```

related failure
```
In [1]: df = pd.DataFrame({
   ...:     'a': ['foo', 'bar', None],
   ...:     'b': ['baz', 'quux', None]
   ...: }, dtype='category')
   ...: df
   ...: 
Out[1]: 
     a     b
0  foo   baz
1  bar  quux
2  NaN   NaN

In [2]: df.where(df.isnull(), None)
---------------------------------------------------------------------------
ValueError: Wrong number of items passed 3, placement implies 1
```
#### Problem description

Object dataframes can contain any Python object so anything should be convertible to them. Split off of #16979.

### LINKED ISSUE #24077: Support ExtensionArray types in where
This is blocking DatetimeArray. It's also a slight regression from 0.24, since things like `.where` on a DataFrame with period objects would work (via object dtype).

I think the easiest place for this is by defining `ExtensionBlock.where`, and restricting it to cases where the dtype of `self` and `other` match (so that the result dtype is the same).

We can do this pretty easily for our EAs by performing the `.where` on `_ndarray_values`. But `_ndarray_values` isn't part of the EA interface yet. I'm not sure if we'll have time to properly design and implement a generic `.where` for any ExtensionArray since there are a couple subtlies.

Here's a start

```diff
diff --git a/pandas/core/internals/blocks.py b/pandas/core/internals/blocks.py
index 1b67c2053..ce5c01359 100644
--- a/pandas/core/internals/blocks.py
+++ b/pandas/core/internals/blocks.py
@@ -1955,6 +1955,37 @@ class ExtensionBlock(NonConsolidatableMixIn, Block):
                                            placement=self.mgr_locs,
                                            ndim=self.ndim)]
 
+    def where(self, other, cond, align=True, errors='raise',
+              try_cast=False, axis=0, transpose=False):
+        import pandas.core.computation.expressions as expressions
+
+        values = self.values._ndarray_values
+
+        if cond.ndim == 2:
+            assert cond.shape[-1] == 1
+            cond = cond._data.blocks[0].values.ravel()
+
+        if hasattr(other, 'ndim') and other.ndim == 2:
+            # TODO: this hasn't been normalized
+            assert other.shape[-1] == 1
+            other = other._data.blocks[0].values
+
+        elif (lib.is_scalar(other) and isna(other)) or other is None:
+            # TODO: we need the storage NA value (e.g. iNaT)
+            other = self.values.dtype.na_value
+            # other = tslibs.iNaT
+
+        # TODO: cond.ravel().all() short-circut
+
+        if cond.ndim > 1:
+            cond = cond.ravel()
+
+        result = expressions.where(cond, values, other)
+        if not isinstance(result, self._holder):
+            # Need a kind of _from_ndarray_values()
+            # this is different from _from_sequence
+            result = self._holder.(result, dtype=self.dtype)
+        return self.make_block_same_class(result)
+
     @property
     def _ftype(self):
         return getattr(self.values, '_pandas_ftype', Block._ftype)
```

There are a couple TODOs there, plus tests, and I'm sure plenty of edge cases.

```python
In [7]: df = pd.DataFrame({"A": pd.period_range("2000", periods=12)})

In [8]: df.where(df.A.dt.day == 2)
Out[8]:
             A
0          NaT
1   2000-01-02
2          NaT
3          NaT
4          NaT
5          NaT
6          NaT
7          NaT
8          NaT
9          NaT
10         NaT
11         NaT
```

### LINKED ISSUE #24142: Categorical.__setitem__ should allow setting a Categorical with same unordered categories
We'd need to recode the source categorical's codes to the target's.

```python
In [6]: a = pd.Categorical(['a', 'b'], categories=['b', 'a'])

In [7]: b = pd.Categorical(['a', 'b'], categories=['a', 'b'])

In [8]: a[[True, False]] = b[[True, False]]
```

```pytb
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-8-9693ad7c05cc> in <module>
----> 1 a[[True, False]] = b[[True, False]]

~/sandbox/pandas/pandas/core/arrays/categorical.py in __setitem__(self, key, value)
   2082         if isinstance(value, Categorical):
   2083             if not value.categories.equals(self.categories):
-> 2084                 raise ValueError("Cannot set a Categorical with another, "
   2085                                  "without identical categories")
   2086

ValueError: Cannot set a Categorical with another, without identical categories
```

### PR DESCRIPTION
**Pull request by TomAugspurger**
We need some way to do `.where` on EA object for DatetimeArray. Adding it
to the interface is, I think, the easiest way.

Initially I started to write a version on ExtensionBlock, but it proved
to be unwieldy. to write a version that performed well for all types.
It *may* be possible to do using `_ndarray_values` but we'd need a few more
things around that (missing values, converting an arbitrary array to the
"same' ndarary_values, error handling, re-constructing). It seemed easier
to push this down to the array.

The implementation on ExtensionArray is readable, but likely slow since
it'll involve a conversion to object-dtype.

Closes #24077
Closes #24142
Closes #16983

### CODE REVIEW

File: doc/source/whatsnew/v0.24.0.rst [hunk 1]
```diff
@@ -1262,6 +1264,7 @@ Categorical
 - In meth:`Series.unstack`, specifying a ``fill_value`` not present in the categories now raises a ``TypeError`` rather than ignoring the ``fill_value`` (:issue:`23284`)
 - Bug when resampling :meth:`Dataframe.resample()` and aggregating on categorical data, the categorical dtype was getting lost. (:issue:`23227`)
 - Bug in many methods of the ``.str``-accessor, which always failed on calling the ``CategoricalIndex.str`` constructor (:issue:`23555`, :issue:`23556`)
+- Bug in :meth:`Series.where` losing the categorical dtype for categorical data (:issue:`24077`)
```
[the remark this answers is not in the document]
Comment (jreback) [reply doc/source/whatsnew/v0.24.0.rst]: if its in the categories then this should work and return categorical, if its not hen i think coercing to object is ok

File: pandas/core/arrays/base.py [hunk 2]
```diff
@@ -661,6 +662,40 @@ def take(self, indices, allow_fill=False, fill_value=None):
...
+        -------
+        ExtensionArray
+            Same dtype as the original.
+
+        See Also
+        --------
+        Series.where : Similar method for Series.
+        DataFrame.where : Similar method for DataFrame.
+        """
+        return type(self)._from_sequence(np.where(cond, self, other),
```
Comment (jreback) [inline pandas/core/arrays/base.py]: hmm this turns it into an array. we have much special handling for this (e.g. see .where for DTI). i think this needs to dispatch somehow.

Comment (jreback) [reply pandas/core/arrays/base.py]: oh I see you override things. ok then.

**Comment by TomAugspurger** [comment]
Hmm I don't like the return dtype depending on the values.

Perhaps we do this with a deprecation warning?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by TomAugspurger** [comment]
@jorisvandenbossche do you have any objections to adding `ExtensionArray.where` to the interface? Or do the behavior change on `Series[category].where`?

**Comment by jorisvandenbossche** [comment]
Just for context: how is this different from `eaarray[cond] = other` ?

The behaviour change to keep the categorical dtype is certainly fine.

**Comment by TomAugspurger** [comment]
I suppose that `np.where` would work on items that don't implement
`__setitem__`. But other than that they should be identical
for 1-d arrays, right?

**Comment by jorisvandenbossche** [comment]
> But other than that they should be identical for 1-d arrays, right?

And since EAs are 1D, and our internal EAs support setitem, why is the new code needed? Or what in setitem is not working as it should right now? (maybe I am missing some context)

**Comment by TomAugspurger** [comment]
This came out  of the DatetimeArray refactor. I'll have to take another
look at exactly what the failures where. They were pretty deep in the
internals.

File: pandas/core/arrays/base.py [hunk 3]
```diff
@@ -661,6 +662,42 @@ def take(self, indices, allow_fill=False, fill_value=None):
         # pandas.api.extensions.take
         raise AbstractMethodError(self)
 
+    def where(self, cond, other):
```
Comment (jschendel) [inline pandas/core/arrays/base.py]: The other implementations of `where` (`DataFrame.where`, `Index.where`, etc.) have `other` default to NA.  Do we want to maintain that convention here too?

File: pandas/core/arrays/interval.py [hunk 4]
```diff
@@ -777,6 +777,17 @@ def take(self, indices, allow_fill=False, fill_value=None, axis=None,
...
 
+    def where(self, cond, other):
+        if is_scalar(other) and isna(other):
+            lother = rother = other
+        else:
+            self._check_closed_matches(other, name='other')
+            lother = other.left
+            rother = other.right
+        left = np.where(cond, self.left, lother)
+        right = np.where(cond, self.right, rother)
```
Comment (jschendel) [inline pandas/core/arrays/interval.py]: `left`/`right` should have a `where` method, so might be a bit safer to do something like:
```python
left = self.left.where(cond, lother)
right = self.right.where(cond, rother)
```
`np.where` looks like it can cause some problems depending on what `left`/`right` are:
```python
In [2]: left = pd.date_range('2018', periods=3); left
Out[2]: DatetimeIndex(['2018-01-01', '2018-01-02', '2018-01-03'], dtype='datetime64[ns]', freq='D')

In [3]: np.where([True, False, True], left, pd.NaT)
Out[3]: array([1514764800000000000, NaT, 1514937600000000000], dtype=object)
```

File: pandas/core/arrays/interval.py [hunk 5]
```diff
@@ -777,6 +777,17 @@ def take(self, indices, allow_fill=False, fill_value=None, axis=None,
 
         return self._shallow_copy(left_take, right_take)
 
+    def where(self, cond, other):
```
Comment (jschendel) [inline pandas/core/arrays/interval.py]: Would be nice to have `IntervalIndex` use this implementation instead of the naive object array based implementation that it currently uses.  Can certainly leave that for a follow-up PR though, and I'd be happy to do it.

File: pandas/core/arrays/interval.py [hunk 6]
```diff
@@ -777,6 +777,17 @@ def take(self, indices, allow_fill=False, fill_value=None, axis=None,
...
+        if is_scalar(other) and isna(other):
+            lother = rother = other
```
Comment (jschendel) [inline pandas/core/arrays/interval.py]: To be safe, I think this should be `lother = rother = self.left._na_value` to ensure that we're filling `left`/`right` with the correct NA value.  If we use `left/right.where` instead of `np.where` this _should_ be handled automatically iirc, so could maybe just do that instead.

File: pandas/core/arrays/interval.py [hunk 7]
```diff
@@ -777,6 +777,17 @@ def take(self, indices, allow_fill=False, fill_value=None, axis=None,
...
+        else:
```
Comment (jschendel) [inline pandas/core/arrays/interval.py]: Can you make this an `elif` that checks that `other` is interval-like (something like `isinstance(other, Interval) or is_interval_dtype(other)`), then have an `else` clause that raises a `ValueError` saying `other` must be interval-like?  

As written I think this would raise a somewhat unclear `AttributeError` in `self._check_closed_matches` since it assumes `other.closed` exists.

**Comment by TomAugspurger** [comment]
On further reflection, I realize that ndarrays don't have a `where` method, so I don't think we should add `ExtensionArray.where`.

I'll see if setitem on a copy is sufficient.

**Comment by TomAugspurger** [comment]
Updated to use setitem...

Sorry, this kind of grew in scope. I'm going to rebase my datetimearray PR and see which subset is actually needed to fix DTA.

File: pandas/core/indexes/category.py [hunk 8]
```diff
@@ -501,10 +501,13 @@ def _can_reindex(self, indexer):
 
     @Appender(_index_shared_docs['where'])
     def where(self, cond, other=None):
+        # TODO: Investigate an alternative implementation with
```
Comment (TomAugspurger) [inline pandas/core/indexes/category.py]: https://github.com/pandas-dev/pandas/issues/24144

File: pandas/core/internals/blocks.py [hunk 9]
```diff
@@ -1967,6 +1968,57 @@ def shift(self, periods, axis=0):
...
+        if isinstance(cond, (ABCIndexClass, ABCSeries)):
+            cond = cond.array
+
+        if lib.is_scalar(other) and isna(other):
+            # The default `other` for Series / Frame is np.nan
+            # we want to replace that with the correct NA value
+            # for the type
+            other = self.dtype.na_value
+
+        if is_sparse(self.values):
```
Comment (TomAugspurger) [inline pandas/core/internals/blocks.py]: Without this, we fail in the

```python
            result = self._holder._from_sequence(
                 np.where(cond, self.values, other),
                 dtype=dtype,
```

since the `where` may change the dtype, if NaN is introduced.


Implementing `SparseArray.__setitem__` would allow us to remove this block.

**Comment by TomAugspurger** [comment]
> Or what in setitem is not working as it should right now? (maybe I am missing some context)

@jorisvandenbossche OK, here's some context :) The most immediate failure is mismatched block dimensions / shapes for in `DataFrame.where` for EA (not just DatetimeArray, but that was tested). This uses Categorical

```python
In [8]: df = pd.DataFrame({"A": pd.Categorical([1, 2, 3])})

In [9]: df.where(pd.DataFrame({"A": [True, False, True]}))
```

```pytb
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-9-56dcebf7e672> in <module>
----> 1 df.where(pd.DataFrame({"A": [True, False, True]}))

... 

~/sandbox/pandas/pandas/core/internals/blocks.py in __init__(self, values, placement, ndim)
     84             raise ValueError(
     85                 'Wrong number of items passed {val}, placement implies '
---> 86                 '{mgr}'.format(val=len(self.values), mgr=len(self.mgr_locs)))
     87
     88     def _check_ndim(self, values, ndim):

ValueError: Wrong number of items passed 3, placement implies 1
```

The broadcasting is all messed up since the shapes aren't right (we're using `Block.where`).

```python
ipdb> cond
array([[ True],
       [False],
       [ True]])
ipdb> values
[1, 2, 3]
Categories (3, int64): [1, 2, 3]
ipdb> other
nan
```

A hacky, but shorter fix is to use the following (this is in Block.where)

```diff
diff --git a/pandas/core/internals/blocks.py b/pandas/core/internals/blocks.py
index 618b9eb12..2356a226d 100644
--- a/pandas/core/internals/blocks.py
+++ b/pandas/core/internals/blocks.py
@@ -1319,12 +1319,20 @@ class Block(PandasObject):
 
         values = self.values
         orig_other = other
+        if not self._can_consolidate:
+            transpose = False
+
         if transpose:
             values = values.T
 
         other = getattr(other, '_values', getattr(other, 'values', other))
         cond = getattr(cond, 'values', cond)
 
+        if not self._can_consolidate:
+            if cond.ndim == 2:
+                assert cond.shape[-1] == 1
+                cond = cond.ravel()
+
         # If the default broadcasting would go in the wrong direction, then
         # explicitly reshape other instead
         if getattr(other, 'ndim', 0) >= 1:
```

That fixes most of the issues I was having on the DTA branch. Still running the tests to see if any were re-broken.

---

So, in summary

1. we need to support EAs in DataFrame.where
2. The diff just above is a hacky way of achieving just that (no more).
3. This PR will still be useful for avoiding the conversion to object (DatetimeTZBlock will avoid it even without this PR, thanks to `_try_coerce_args` coercing datetimes to ints)

File: pandas/core/internals/blocks.py [hunk 10]
```diff
@@ -1967,6 +1968,57 @@ def shift(self, periods, axis=0):
...
+            other = other.array
+
+        if isinstance(cond, ABCDataFrame):
+            assert cond.shape[1] == 1
+            cond = cond.iloc[:, 0].array
+
+        if isinstance(other, ABCDataFrame):
+            assert other.shape[1] == 1
+            other = other.iloc[:, 0].array
+
```
Comment (jreback) [inline pandas/core/internals/blocks.py]: can you  do the others first, then the conds; you can use if/elif

File: pandas/core/internals/blocks.py (hunk 9 above)
Comment (jreback) [reply pandas/core/internals/blocks.py]: this should be an overriding method in Sparse then, not here

File: pandas/core/internals/blocks.py [hunk 11]
```diff
@@ -1967,6 +1968,57 @@ def shift(self, periods, axis=0):
...
+                otype = other.dtype.subtype
+            else:
+                otype = other
+            dtype = self.dtype.update_dtype(
+                np.result_type(self.values.dtype.subtype, otype)
+            )
+        else:
+            dtype = self.dtype
+
+        # rough heuristic to see if the other array implements setitem
```
Comment (jreback) [inline pandas/core/internals/blocks.py]: again you don't actually need to do this here, rather override in the appropriate class

File: pandas/tests/arrays/categorical/test_indexing.py [hunk 12]
```diff
@@ -122,6 +162,60 @@ def test_get_indexer_non_unique(self, idx_values, key_values, key_class):
             tm.assert_numpy_array_equal(expected, result)
             tm.assert_numpy_array_equal(exp_miss, res_miss)
 
+    def test_where_unobserved_nan(self):
```
Comment (jreback) [inline pandas/tests/arrays/categorical/test_indexing.py]: is where all of the where tests are?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/internals/blocks.py (hunk 9 above)
Comment (TomAugspurger) [reply pandas/core/internals/blocks.py]: We don't have a SparseBlock anymore. I can add one back if you want, but I figured it'd be easier not to since implementing `SparseArray.__setitem__` will remove the need for this, and we'd just have to remove SparseBlock again.

File: pandas/core/internals/blocks.py (hunk 11 above)
Comment (TomAugspurger) [reply pandas/core/internals/blocks.py]: We will still need the check for extension, even if we create SparseBlock again.

File: pandas/tests/arrays/categorical/test_indexing.py (hunk 12 above)
Comment (TomAugspurger) [reply pandas/tests/arrays/categorical/test_indexing.py]: There weren't any previously (we used to fall back to object).

**Comment by TomAugspurger** [comment]
Updated. Main outstanding point is whether or not we should create a SparseBlock just for this. I don't have a preference.

File: pandas/core/internals/blocks.py [hunk 13]
```diff
@@ -1967,6 +1967,57 @@ def shift(self, periods, axis=0):
                                            placement=self.mgr_locs,
                                            ndim=self.ndim)]
 
+    def where(self, other, cond, align=True, errors='raise',
+              try_cast=False, axis=0, transpose=False):
+        if isinstance(other, (ABCIndexClass, ABCSeries)):
+            other = other.array
+
+        elif isinstance(other, ABCDataFrame):
```
Comment (jreback) [inline pandas/core/internals/blocks.py]: can you add some comments here

File: pandas/core/internals/blocks.py (hunk 9 above)
Comment (jreback) [reply pandas/core/internals/blocks.py]: this is pretty hacky. This was why we had originally a ``.get_values()`` methon on Sparse to do things like this. We need something to give back the underlying type of the object, which is useful for Categorical as well. Would rather create a generalized soln than hack it like this.

File: pandas/core/internals/blocks.py [hunk 14]
```diff
@@ -1967,6 +1967,57 @@ def shift(self, periods, axis=0):
...
+                otype = other
+            dtype = self.dtype.update_dtype(
+                np.result_type(self.values.dtype.subtype, otype)
+            )
+        else:
+            dtype = self.dtype
+
+        # rough heuristic to see if the other array implements setitem
+        if (self._holder.__setitem__ == ExtensionArray.__setitem__
+                or self._holder.__setitem__ == SparseArray.__setitem__):
```
Comment (jreback) [inline pandas/core/internals/blocks.py]: what the heck is this?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/internals/blocks.py (hunk 9 above)
Comment (TomAugspurger) [reply pandas/core/internals/blocks.py]: Actually, we don't need this. I think we can just re-infer the dtype from the output of `np.where`.

File: pandas/core/internals/blocks.py (hunk 14 above)
Comment (TomAugspurger) [reply pandas/core/internals/blocks.py]: The general block is to check if the block implements `__setitem__`. That specific line is backwards compat for `SparseArray`, which implements `__setitem__` to raise a TypeError instead of a NotImplementedError.

I suppose it'd be cleaner to do this in a `try / except` block...

**Comment by TomAugspurger** [comment]
Cleaned things up a bit I think.

**Comment by TomAugspurger** [comment]
All green.

File: pandas/core/internals/blocks.py (hunk 9 above)
Comment (jreback) [reply pandas/core/internals/blocks.py]: so is this changing?

File: pandas/core/internals/blocks.py [hunk 15]
```diff
@@ -2658,6 +2708,37 @@ def concat_same_type(self, to_concat, placement=None):
...
+        try:
+            # Attempt to do preserve categorical dtype.
+            result = super(CategoricalBlock, self).where(
+                other, cond, align, errors, try_cast, axis, transpose
+            )
+        except (TypeError, ValueError):
+            if lib.is_scalar(other):
+                msg = object_msg.format(other)
+            else:
+                msg = compat.reprlib.repr(other)
```
Comment (jreback) [inline pandas/core/internals/blocks.py]: why is this needed?

**Review by jreback — CHANGES_REQUESTED** [review]
looks pretty reasonable. question about the sparse checks.

File: pandas/core/internals/blocks.py (hunk 9 above)
Comment (TomAugspurger) [reply pandas/core/internals/blocks.py]: Changing from master? Yes, in the sense that it'll return a SparseArray. But it still densifies when `np.where` is called.

If you mean "is this changing in the future", yes it'll be removed when `SparseArray.__setitem__` is implemented.

File: pandas/core/internals/blocks.py (hunk 15 above)
Comment (TomAugspurger) [reply pandas/core/internals/blocks.py]: So we don't blow up with a long message for large categoricals. I messed it up though, one sec.

Comment (jreback) [reply pandas/core/internals/blocks.py]: k cool

File: pandas/tests/extension/test_categorical.py [hunk 16]
```diff
@@ -25,7 +25,12 @@
 
 
 def make_data():
-    return np.random.choice(list(string.ascii_letters), size=100)
+    while True:
+        values = np.random.choice(list(string.ascii_letters), size=100)
+        # ensure we meet the requirement
```
Comment (jreback) [inline pandas/tests/extension/test_categorical.py]: no repeated values but duplicates allowed?

File: pandas/tests/extension/conftest.py [hunk 17]
```diff
@@ -11,7 +11,11 @@ def dtype():
 
 @pytest.fixture
 def data():
-    """Length-100 array for this type."""
+    """Length-100 array for this type.
```
Comment (jreback) [inline pandas/tests/extension/conftest.py]: can you copy this doc-string to the categorical one

File: pandas/core/internals/blocks.py [hunk 18]
```diff
@@ -2658,6 +2708,32 @@ def concat_same_type(self, to_concat, placement=None):
             values, placement=placement or slice(0, len(values), 1),
             ndim=self.ndim)
 
+    def where(self, other, cond, align=True, errors='raise',
+              try_cast=False, axis=0, transpose=False):
+        # This can all be deleted in favor of ExtensionBlock.where once
```
Comment (jreback) [inline pandas/core/internals/blocks.py]: can you add TODO(EA) or someting here so we know to remove this

File: pandas/core/internals/blocks.py (hunk 9 above)
Comment (jreback) [reply pandas/core/internals/blocks.py]: oh ok, can you add a TODO comment

**Review by jreback — APPROVED** [review]
small additional comments, lgtm otherwise. ping on green.

File: pandas/tests/extension/test_categorical.py (hunk 16 above)
Comment (TomAugspurger) [reply pandas/tests/extension/test_categorical.py]: Just that the first two are distinct., since the `where` test requires that `data[0] != data[1]`.

**Comment by TomAugspurger** [comment]
All green.
